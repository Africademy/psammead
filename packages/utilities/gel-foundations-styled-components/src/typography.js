import { css } from 'styled-components';
import { MEDIA_QUERY_TYPOGRAPHY } from '@bbc/gel-constants/breakpoints';

export const GEL_ATLAS = css`
  font-size: 4.875em;
  line-height: 5.25rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 6em;
    line-height: 6.5rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 8.75em;
    line-height: 9.25rem;
  }
`;

export const GEL_ELEPHANT = css`
  font-size: 3.75em;
  line-height: 4rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 4.875em;
    line-height: 5.25rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 7.25em;
    line-height: 7.75rem;
  }
`;

export const GEL_IMPERIAL = css`
  font-size: 3.125em;
  line-height: 3.375rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 4em;
    line-height: 4.5rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 6em;
    line-height: 6.5rem;
  }
`;

export const GEL_ROYAL = css`
  font-size: 2.5em;
  line-height: 2.75rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 3.25em;
    line-height: 3.75rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 4.75em;
    line-height: 5.25rem;
  }
`;

export const GEL_FOOLSCAP = css`
  font-size: 2em;
  line-height: 2.25rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 2.5em;
    line-height: 2.75rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 3.5em;
    line-height: 3.75rem;
  }
`;

export const GEL_CANON = css`
  font-size: 1.75em;
  line-height: 2rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 2em;
    line-height: 2.25rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 2.75em;
    line-height: 3rem;
  }
`;

export const GEL_TRAFALGAR = css`
  font-size: 1.25em;
  line-height: 1.5rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 1.5em;
    line-height: 1.75rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 2em;
    line-height: 2.25rem;
  }
`;

export const GEL_PARAGON = css`
  font-size: 1.25em;
  line-height: 1.5rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    font-size: 1.375em;
    line-height: 1.625rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 1.75em;
    line-height: 2rem;
  }
`;

export const GEL_DOUBLE_PICA = css`
  font-size: 1.25em;
  line-height: 1.5rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    line-height: 1.5rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 1.5em;
    line-height: 1.75rem;
  }
`;

export const GEL_GREAT_PRIMER = css`
  font-size: 1.125em;
  line-height: 1.375rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    line-height: 1.375rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 1.25em;
    line-height: 1.5rem;
  }
`;

export const GEL_BODY_COPY = css`
  font-size: 0.9375em;
  line-height: 1.25rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_AND_LARGER} {
    font-size: 1em;
    line-height: 1.375rem;
  }
`;

export const GEL_PICA = css`
  font-size: 0.9375em;
  line-height: 1.25rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_AND_LARGER} {
    font-size: 1em;
    line-height: 1.25rem;
  }
`;

export const GEL_LONG_PRIMER = css`
  font-size: 0.9375em;
  line-height: 1.125rem;

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 0.875em;
  }
`;

export const GEL_BREVIER = css`
  font-size: 0.875em;
  line-height: 1rem;

  ${MEDIA_QUERY_TYPOGRAPHY.SMART_PHONE_ONLY} {
    line-height: 1.125rem;
  }

  ${MEDIA_QUERY_TYPOGRAPHY.LAPTOP_AND_LARGER} {
    font-size: 0.8125em;
  }
`;

export const GEL_MINION = css`
  font-size: 0.75em;
  line-height: 1em;
`;
