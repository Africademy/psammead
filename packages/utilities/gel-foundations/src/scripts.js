/* eslint-disable global-require */
module.exports = {
  latin: require('../dist/scripts/latin-and-cyrillic').default,
  cyrillic: require('../dist/scripts/latin-and-cyrillic').default,
  latinDiacritics: require('../dist/scripts/latin-with-diacritics').default,
  arabic: require('../dist/scripts/arabic').default,
  arabicPashto: require('../dist/scripts/arabic-pashto').default,
  bengali: require('../dist/scripts/bengali').default,
  burmese: require('../dist/scripts/burmese').default,
  hindi: require('../dist/scripts/devanagari-and-gurmukhi').default,
  nepali: require('../dist/scripts/devanagari-and-gurmukhi').default,
  sinhalese: require('../dist/scripts/sinhalese').default,
  tamil: require('../dist/scripts/tamil').default,
  thai: require('../dist/scripts/thai').default,
  chinese: require('../dist/scripts/no-ascenders-or-descenders').default,
  korean: require('../dist/scripts/no-ascenders-or-descenders').default,
  ethiopic: require('../dist/scripts/no-ascenders-or-descenders').default,
};
