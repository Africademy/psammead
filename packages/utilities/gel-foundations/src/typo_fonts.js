const fontFamilyBase = ', Helvetica, Arial, sans-serif';
const baseFontUrl = 'https://gel.files.bbci.co.uk/r2.302/';

export const FF_SERIF = `ReithSerif${fontFamilyBase}`;
export const FF_SANS = `ReithSans${fontFamilyBase}`;
export const FF_SANS_COND = `ReithSansCondensed${fontFamilyBase}`;

// Serif
export const T_REITH_SERIF_REGULAR = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_Rg.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_Rg.woff") format("woff");
    font-display: optional;
  }`;

export const T_REITH_SERIF_ITALIC = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_It.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_It.woff") format("woff");
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SERIF_BOLD = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_Bd.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_Bd.woff") format("woff");
    font-weight: bold;
    font-display: optional;
  }`;

export const T_REITH_SERIF_BOLD_ITALIC = `  
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_BdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_BdIt.woff") format("woff");
    font-weight: bold;
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SERIF_LIGHT = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_Lt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_Lt.woff") format("woff");
    font-weight: 300;
    font-display: optional;
  }`;

export const T_REITH_SERIF_LIGHT_ITALIC = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_LtIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_LtIt.woff") format("woff");
    font-weight: 300;
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SERIF_MEDIUM = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_Md.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_Md.woff") format("woff");
    font-weight: 500;
    font-display: optional;
  }`;

export const T_REITH_SERIF_MEDIUM_ITALIC = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_MdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_MdIt.woff") format("woff");
    font-weight: 500;
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SERIF_EXTRA_BOLD = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_ExBd.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_ExBd.woff") format("woff");
    font-weight: 800;
    font-display: optional;
  }`;

export const T_REITH_SERIF_EXTRA_BOLD_ITALIC = `
  @font-face {
    font-family: "ReithSerif";
    src: url("${baseFontUrl}BBCReithSerif_W_ExBdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSerif_W_ExBdIt.woff") format("woff");
    font-weight: 800;
    font-style: italic;
    font-display: optional;
  }`;

// Sans
export const T_REITH_SANS_REGULAR = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_Rg.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_Rg.woff") format("woff");
    font-display: optional;
  }`;

export const T_REITH_SANS_ITALIC = `
    @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_It.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_It.woff") format("woff");
    font-style: italic;
    font-display: optional;
  }
`;

export const T_REITH_SANS_BOLD = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_Bd.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_Bd.woff") format("woff");
    font-weight: bold;
    font-display: optional;
  }
  `;

export const T_REITH_SANS_BOLD_ITALIC = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_BdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_BdIt.woff") format("woff");
    font-weight: bold;
    font-style: italic;
    font-display: optional;
  }
  `;

export const T_REITH_SANS_LIGHT = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_Lt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_Lt.woff") format("woff");
    font-weight: 300;
    font-display: optional;
  }`;

export const T_REITH_SANS_LIGHT_ITALIC = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_LtIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_LtIt.woff") format("woff");
    font-weight: 300;
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SANS_MEDIUM = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_Md.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_Md.woff") format("woff");
    font-weight: 500;
    font-display: optional;
  }`;

export const T_REITH_SANS_MEDIUM_ITALIC = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_MdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_MdIt.woff") format("woff");
    font-weight: 500;
    font-style: italic;
    font-display: optional;
  }`;

export const T_REITH_SANS_EXTRA_BOLD = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_ExBd.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_ExBd.woff") format("woff");
    font-weight: 800;
    font-display: optional;
  }`;

export const T_REITH_SANS_EXTRA_BOLD_ITALIC = `
  @font-face {
    font-family: "ReithSans";
    src: url("${baseFontUrl}BBCReithSans_W_ExBdIt.woff2") format("woff2"), url("${baseFontUrl}BBCReithSans_W_ExBdIt.woff") format("woff");
    font-weight: 800;
    font-style: italic;
    font-display: optional;
  }`;

// Sans Condensed
export const T_REITH_SANS_CONDENSED_REGULAR = `
    @font-face {
        font-family: "ReithSansCondensed";
        src: url("${baseFontUrl}BBCReithSansCd_W_Rg.woff2") format("woff2"), url("${baseFontUrl}BBCReithSansCd_W_Rg.woff") format("woff");
        font-display: optional;
    }`;

export const T_REITH_SANS_CONDENSED_BOLD = `
    @font-face {
        font-family: "ReithSansCondensed";
        font-weight: bold;
        src: url("${baseFontUrl}BBCReithSansCd_W_Bd.woff2") format("woff2"), url("${baseFontUrl}BBCReithSansCd_W_Bd.woff") format("woff");
        font-display: optional;
    }`;
