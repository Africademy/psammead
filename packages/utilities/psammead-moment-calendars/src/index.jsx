import React from 'react';

const MomentCalendars = () => <h1>Hello World</h1>;

export default MomentCalendars;
