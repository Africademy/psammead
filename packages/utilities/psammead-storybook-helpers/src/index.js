import inputProvider from './input-provider';
import { dirDecorator } from './decorators';
import LANGUAGE_VARIANTS from './text-variants';

export { inputProvider, LANGUAGE_VARIANTS, dirDecorator };
