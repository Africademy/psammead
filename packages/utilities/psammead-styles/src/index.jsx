import { shape, func } from 'prop-types';
import styled from 'styled-components';
import { scriptPropType } from '@bbc/gel-foundations/prop-types';

const TypographyText = styled.p`
  ${props =>
    props.script && props.typographyFunc
      ? props.typographyFunc(props.script)
      : ''}
`;

TypographyText.propTypes = {
  script: shape(scriptPropType).isRequired,
  typographyFunc: func.isRequired,
};

export default TypographyText;
