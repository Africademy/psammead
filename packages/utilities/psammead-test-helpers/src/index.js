import { render, waitForDomChange } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';

const renderWithHelmet = async component => {
  /*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:
 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });
 */

  const { container } = render(component);
  const noop = () => {};
  const ARBITRARY_TIMEOUT = 100; // seems long enough for any dom mutations to occur

  return waitForDomChange({
    container: document.querySelector('head'),
    timeout: ARBITRARY_TIMEOUT,
  })
    .catch(noop) // handle a waitForDomChange timeout
    .then(mutationsList => {
      const domMutationDetected = mutationsList && mutationsList.length;

      if (domMutationDetected) {
        // helmet was probably used so we should get the full html

        const htmlElement = document.querySelector('html');

        const helmetElements = document.querySelectorAll(
          '[data-react-helmet="true"]',
        );

        const removeHelmetAttributes = el =>
          el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

        removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

        Array.from(helmetElements).forEach(removeHelmetAttributes);

        return { container: document.querySelector('html') };
      }

      return { container };
    });
};

export const shouldMatchSnapshot = (title, component) => {
  it(title, async () => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    const { container } = await renderWithHelmet(component);
    const hasOneChild = container.children.length === 1;
    /*
     * if the container has more than one child then it's a component that uses a
     * fragment at the top level so we should not select the first child because it
     * wouldn't snapshot the whole component
     */
    expect(
      hasOneChild ? removeWrappingDiv(container) : container,
    ).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
