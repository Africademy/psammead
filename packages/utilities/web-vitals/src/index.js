import fetch from 'cross-fetch';
import { useEffect } from 'react';
import { getCLS, getFID, getLCP, getFCP, getTTFB } from 'web-vitals';
import {
  useNetworkStatus,
  useHardwareConcurrency,
  useMemoryStatus,
} from 'react-adaptive-hooks';
import useEvent from './use-event';

const noOp = () => {};

const webVitalsBase = {
  age: 0,
  type: 'web-vitals',
  url: 'current-page-url',
};

const vitals = { cls: null, fid: null, lcp: null, fcp: null, ttfb: null };
const deviceMetrics = {
  device_mem: null,
  device_cpu: null,
  device_effective_connection: null,
};

const updateWebVitals = ({ name, value }) => {
  const vitalName = name.toLowerCase();
  vitals[vitalName] = value;
};

const updateDeviceMetrics = ({
  deviceMemory,
  numberOfLogicalProcessors,
  effectiveConnectionType,
}) => {
  deviceMetrics.device_mem = deviceMemory;
  deviceMetrics.device_cpu = numberOfLogicalProcessors;
  deviceMetrics.device_effective_connection = effectiveConnectionType;
};

const setCurrentUrl = () => {
  webVitalsBase.url = window.location.href;
};

const appendReportParams = (reportingEndpoint, reportParams) => {
  const url = new URL(reportingEndpoint);
  const paramsString = reportParams
    .map(param => `${param.name}=${param.value}`)
    .join('&');

  return url.search
    ? `${reportingEndpoint}&${paramsString}`
    : `${reportingEndpoint}?${paramsString}`;
};

const sendBeacon = (rawBeacon, reportingEndpoint, reportParams) => {
  const beacon = JSON.stringify(rawBeacon);
  const beaconTarget = reportParams.length
    ? appendReportParams(reportingEndpoint, reportParams)
    : reportingEndpoint;

  if (navigator.sendBeacon) {
    return navigator.sendBeacon(beaconTarget, beacon);
  }
  return fetch(beaconTarget, {
    method: 'POST',
    headers: { 'Content-Type': 'application/reports+json' },
    body: beacon,
  });
};

const shouldSample = sampleRate => {
  const randomNumber = Math.floor(Math.random() * 100);
  return randomNumber <= sampleRate;
};

const useWebVitals = ({
  enabled,
  reportingEndpoint,
  loggerCallback = noOp,
  sampleRate = 100,
  reportParams = [],
}) => {
  let pageLoadTime;
  const { effectiveConnectionType } = useNetworkStatus();
  const { numberOfLogicalProcessors } = useHardwareConcurrency();
  const { deviceMemory } = useMemoryStatus();

  const sendVitals = async () => {
    const pageExitTime = Date.now();
    const pageAge = pageExitTime - pageLoadTime;

    const beacon = [
      { ...webVitalsBase, age: pageAge, body: { ...vitals, ...deviceMetrics } },
    ];

    try {
      if (shouldSample(sampleRate)) {
        await sendBeacon(beacon, reportingEndpoint, reportParams);
      }
    } catch (error) {
      loggerCallback(error);
    }
  };

  useEvent('pagehide', enabled ? sendVitals : noOp);

  useEffect(() => {
    pageLoadTime = Date.now();
    setCurrentUrl();
    updateDeviceMetrics({
      effectiveConnectionType,
      numberOfLogicalProcessors,
      deviceMemory,
    });
    getCLS(updateWebVitals);
    getFID(updateWebVitals);
    getLCP(updateWebVitals);
    getFCP(updateWebVitals);
    getTTFB(updateWebVitals);
  }, []);

  return null;
};

export default useWebVitals;
