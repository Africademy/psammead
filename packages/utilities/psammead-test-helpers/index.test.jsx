import * as testHelpers from '.';
import * as testHelpersFromSrc from './src/index';

/*
  If you are doing anything than extending this export list it's likely to introduce a breaking change
*/

const testHelpersExpectedExports = {
  shouldMatchSnapshot: 'function',
  shallowRender: 'function',
  shouldShallowMatchSnapshot: 'function',
  isNull: 'function',
  testUtilityPackages: 'function',
};

const expectedExports = { testHelpers: testHelpersExpectedExports };

const actualExports = { testHelpers };

const actualExportsFromSrc = { testHelpers: testHelpersFromSrc };

const testTheTestHelperErrorCases = testHelperMethod => {
  const consoleLogSpy = jest.spyOn(global.console, 'log');

  const actual = { utility: { foo: {} } };
  const expected = { utility: { bar: {} } };

  expect(() => {
    testHelperMethod(actual, expected, 'testing');
  }).toThrowError();
  expect(consoleLogSpy).toHaveBeenCalledWith(
    'No expected export value declared in unit tests for testing/utility.js - foo',
  );

  consoleLogSpy.mockRestore();
};

describe('Psammead test helpers', () => {
  it('should test all the utility exports exist and are the correct type', () => {
    testHelpers.testUtilityPackages(
      actualExports,
      expectedExports,
      'psammead-test-helpers',
    );
  });

  it('should test all the utility exports exist and are the correct type when coming from src/', () => {
    testHelpers.testUtilityPackages(
      actualExportsFromSrc,
      expectedExports,
      'psammead-test-helpers',
    );
  });

  it('should error and console.log if expectedExport is missing a value compared to the actual export', () => {
    testTheTestHelperErrorCases(testHelpers.testUtilityPackages);
  });

  it('should error and console.log if expectedExport is missing a value compared to the actual export when coming from /src', () => {
    testTheTestHelperErrorCases(testHelpersFromSrc.testUtilityPackages);
  });
});
