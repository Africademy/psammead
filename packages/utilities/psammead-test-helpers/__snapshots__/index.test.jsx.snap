// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Psammead test helpers should test all the utility exports exist and are the correct type 1`] = `
<div>
  <h1>
    Hello I am a test component
  </h1>
</div>
`;

exports[`Psammead test helpers should test all the utility exports exist and are the correct type 2`] = `
<html
  dir="rtl"
  lang="fa"
>
  <head>
    <meta
      name="test name"
      value="test value"
    />
    <script
      src="test.js"
    />
  </head>
  <body>
    <div>
      <main>
        <h1>
          Hello I am a test component with React Helmet
        </h1>
      </main>
    </div>
  </body>
</html>
`;
